Various fixes/improvements to parser generation

br0nstein opened the pull request:
This PR includes a few fixes for handling of non terminals and string literals, as well as some tweaks for a cleaner output.

The code change is rather large, as it refactors the serialization of each parser rule into a two-step process. Step 1: recursively convert the javacc expansion to tree representation of ANTLR-domain rules. Step 2: Recursively print the tree nodes. The motivation for doing so is that it allows for extending this project to add depth-based formatting, by recursively collapsing single-element sequences into their underlying single element. This is necessary in order to accurately format the rules differently by depth, a feature my company plans to add in our fork and that others might want to do.

For example, with this new architecture we can easily add a feature to print top-level choices on newlines (and nested choices on the same line), that handles rules like the following:
```
(((<LOREM> | <IPSUM>) { .. java code .. }) ({ .. java code .. } { .. java code .. } ))
```
(The result is `LOREM | IPSUM`)
Previously we printed the rule elements bottom-up, which meant that at the time of processing the `<LOREM> | <IPSUM>` we wouldn't know it is actually top-level, since all its parents are actually single-element sequences once the non-printable expansions are filtered out.

## Fixes

### Emit parser call non terminals that appear within actions

Such parser calls can appear within actions. They need to be treated the same.

```
PARSER_BEGIN(Test)

public class NonTerminalsInActions {}

PARSER_END(SyntaxChecker)

void Choice() : {} {
    <DOLOR> | <SIT>
}

void NonTerminalsInActions() : { NonTerminalsInActions obj; } {
    { Choice(); } { obj = new NonTerminalsInActions(); }
}

```

Previous
```
nonTerminalsInActions :     ;
```

Now
```
nonTerminalsInActions : choice  ;
```

### Don't emit non-BNF non terminals (e.g. JAVACODE productions)

These would be unrecognized by the ANTLR compiler as they aren't handled by this tool and wouldn't correspond to parser/lexer rules. (Prior to 8add27a76c4409252252ec53fc892c18a7b61520 we threw a NPE whenever the source grammar had such productions.)

```
JAVACODE
void E() {}

void NonTerminals() : {} {
    ChoiceEmpty() E() Sequence()
}
```

Previous
```
choiceEmpty : ( e |  e)  ;
nonTerminals :  choiceEmpty e sequence  ;
```

Now
```
choiceEmpty :   ;
nonTerminals : choiceEmpty sequence  ;
```

### Correctly handle string literals

Unlike javacc, in ANTLR all string literals that appear in parser rules must be covered by an existing lexer rule that has a literal body exactly matching. For example, if a parser rule has "HELLO" then there must be a lexer rule RULEXYZ: 'HELLO'. Otherwise, ANTLR grammar compilation would fail. This means, we should throw an exception if there is no matching lexer rule for the string literal, rather than directly emitting it (and previously in the fallback case they were emitted as double quoted literals which ANTLR also doesn't allow). In order to properly handle case insensitive grammar, we emit as the lexer rule name (which would use case-insensitive letter fragments under the hood). Also, we must handle searching across all rule modes for a rule that covers that literal, rather than just default, in case it was only defined in some other state.

```
options {
    IGNORE_CASE = true;
}
<MODE2> TOKEN : {
    <BEGIN: "START" >
}
TOKEN : {
    <END: "FINISH" >
}

void Literal() : {} {
    "FINISH"
}
void LiteralCaseMismatch() : {} {
    "start"
}
void LiteralCaseMatch() : {} {
    "START"
}
```

Previous
```
literal: END  ;
literalCaseMismatch :  "start"  ;   <-- not even valid ANTLR
literalCaseMatch :  "START"  ;   <-- not even valid ANTLR
```

Now
```
literal: END  ;
literalCaseMismatch : BEGIN  ;
literalCaseMatch : BEGIN  ;
```

##  Tweaks

### Convert choices with empty option to be zero-or-one choice of only the non-empty options.

The rule is handled equivalently by ANTLR but is more descriptive to read without the dangling empty choice

```
void ChoiceWithDefaultAction() : {} {
	<DOLOR> | <SIT> | {}
}
```

Previous
```
choiceWithDefaultAction : ( DOLOR |  SIT |  )  ;
```

Now
```
choiceWithDefaultAction : (DOLOR | SIT)?  ;
```

### Don't wrap quantified choices in extra parentheses if already parenthesized

The rule is handled equivalently but is more conventional without redundant parentheses.

```
void ChoiceZeroOrMore() : {} {
    (<DOLOR> | <SIT>)*
}
```

Previous
```
choiceZeroOrMore :  (( DOLOR |  SIT))*  ;
```

Now
```
choiceZeroOrMore : (DOLOR | SIT)*  ;
```

br0nstein committed:
Various fixes to parsing

ftomassetti commented:
Given this code is unmaintained and it looks good at a first look I will proceed to merge it. Thank you for your contribution

br0nstein commented:
Thanks! Just a note that the "Don't wrap quantified choices in extra parentheses if already parenthesized" is not accurate - I forgot I removed that from the PR before uploading as part of removing stuff I thought might be picky to our use case. I have updated the PR description to reflect that.
